// nativescript-plugin-firebase/performance/performance.ios.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var utils_1 = require("../utils");
function startTrace(name) {
    return new FirebaseTrace(FIRPerformance.sharedInstance().traceWithName(name));
}
exports.startTrace = startTrace;
var FirebaseTrace = (function () {
    function FirebaseTrace(nativeTrace) {
        this.nativeTrace = nativeTrace;
    }
    FirebaseTrace.prototype.setValue = function (attribute, value) {
        this.nativeTrace.setValueForAttribute(value, attribute);
    };
    FirebaseTrace.prototype.getValue = function (attribute) {
        return this.nativeTrace.valueForAttribute(attribute);
    };
    FirebaseTrace.prototype.getAttributes = function () {
        return utils_1.firebaseUtils.toJsObject(this.nativeTrace.attributes);
    };
    FirebaseTrace.prototype.removeAttribute = function (attribute) {
        this.nativeTrace.removeAttribute(attribute);
    };
    FirebaseTrace.prototype.incrementMetric = function (metric, by) {
        this.nativeTrace.incrementMetricByInt(metric, by);
    };
    FirebaseTrace.prototype.stop = function () {
        this.nativeTrace.stop();
    };
    return FirebaseTrace;
}());
exports.FirebaseTrace = FirebaseTrace;
