// nativescript-plugin-firebase/mlkit/custommodel/index.ios.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var image_source_1 = require("tns-core-modules/image-source");
var custommodel_common_1 = require("./custommodel-common");
var MLKitCustomModel = (function (_super) {
    __extends(MLKitCustomModel, _super);
    function MLKitCustomModel() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    MLKitCustomModel.prototype.createDetector = function () {
        return getInterpreter();
    };
    MLKitCustomModel.prototype.createSuccessListener = function () {
        var _this = this;
        return function (outputs, error) {
            if (error !== null) {
                console.log(error.localizedDescription);
            }
            else if (outputs !== null) {
                var result = {
                    result: []
                };
                console.log(">>> outputs: " + outputs);
                _this.notify({
                    eventName: MLKitCustomModel.scanResultEvent,
                    object: _this,
                    value: result
                });
            }
        };
    };
    MLKitCustomModel.prototype.rotateRecording = function () {
        return false;
    };
    return MLKitCustomModel;
}(custommodel_common_1.MLKitCustomModel));
exports.MLKitCustomModel = MLKitCustomModel;
function getInterpreter() {
    var fIRModelDownloadConditions = FIRModelDownloadConditions.alloc().initWithIsWiFiRequiredCanDownloadInBackground(false, true);
    var fIRCloudModelSource = FIRCloudModelSource.alloc().initWithModelNameEnableModelUpdatesInitialConditionsUpdateConditions("my-custom-model", true, fIRModelDownloadConditions, fIRModelDownloadConditions);
    loadLocalModel();
    var fIRModelOptions = FIRModelOptions.alloc().initWithCloudModelNameLocalModelName(null, "mobilenet");
    return FIRModelInterpreter.modelInterpreterWithOptions(fIRModelOptions);
}
function loadLocalModel() {
    var localModelFilePath = NSBundle.mainBundle.pathForResourceOfType("mobilenet_quant_v1_224", "tflite");
    console.log(">>> localModelFilePath: " + localModelFilePath);
    var localModelSource = FIRLocalModelSource.alloc().initWithModelNamePath("mobilenet", localModelFilePath);
    console.log(">>> localModelSource: " + localModelSource);
    var localModelRegistrationSuccess = FIRModelManager.modelManager().registerLocalModelSource(localModelSource);
    console.log("localModelRegistrationSuccess: " + localModelRegistrationSuccess);
}
function useCustomModel(options) {
    return new Promise(function (resolve, reject) {
        try {
            var modelInterpreter = getInterpreter();
            var inputs = FIRModelInputs.new();
            var image = options.image instanceof image_source_1.ImageSource ? options.image.ios : options.image.imageSource.ios;
            var resizedImg = TNSMLKitCameraView.resizeImage(image);
            var successAddingInput = inputs.addInputError(resizedImg);
            var inputOptions = FIRModelInputOutputOptions.new();
            var arrIn = NSMutableArray.new();
            arrIn.addObject(1);
            arrIn.addObject(image.size.width);
            arrIn.addObject(image.size.height);
            arrIn.addObject(3);
            var arrOut = NSMutableArray.new();
            arrOut.addObject(1);
            arrOut.addObject(1001);
            inputOptions.setInputFormatForIndexTypeDimensionsError(0, 3, arrIn);
            inputOptions.setOutputFormatForIndexTypeDimensionsError(0, 3, arrOut);
            modelInterpreter.runWithInputsOptionsCompletion(inputs, inputOptions, function (outputs, error) {
                console.log(">>> error: " + error);
                console.log(">>> outputs: " + outputs);
                if (error !== null) {
                    reject(error.localizedDescription);
                }
                else if (outputs !== null) {
                    console.log(">>> outputs.count: " + outputs.outputAtIndexError(0));
                    var result = {
                        result: []
                    };
                    console.log(">>> outputs: " + outputs);
                    resolve(result);
                }
            });
        }
        catch (ex) {
            console.log("Error in firebase.mlkit.useCustomModel: " + ex);
            reject(ex);
        }
    });
}
exports.useCustomModel = useCustomModel;
