// nativescript-plugin-firebase/mlkit/barcodescanning/index.ios.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var image_source_1 = require("tns-core-modules/image-source");
var barcodescanning_common_1 = require("./barcodescanning-common");
exports.BarcodeFormat = barcodescanning_common_1.BarcodeFormat;
var MLKitBarcodeScanner = (function (_super) {
    __extends(MLKitBarcodeScanner, _super);
    function MLKitBarcodeScanner() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    MLKitBarcodeScanner.prototype.createDetector = function () {
        var formats;
        if (this.formats) {
            formats = [];
            var requestedFormats = this.formats.split(",");
            requestedFormats.forEach(function (format) { return formats.push(barcodescanning_common_1.BarcodeFormat[format.trim().toUpperCase()]); });
        }
        return getBarcodeDetector(formats);
    };
    MLKitBarcodeScanner.prototype.createSuccessListener = function () {
        var _this = this;
        return function (barcodes, error) {
            if (error !== null) {
                console.log(error.localizedDescription);
            }
            else if (barcodes !== null) {
                var result = {
                    barcodes: []
                };
                for (var i = 0, l = barcodes.count; i < l; i++) {
                    var barcode = barcodes.objectAtIndex(i);
                    result.barcodes.push({
                        value: barcode.rawValue,
                        format: barcodescanning_common_1.BarcodeFormat[barcode.format],
                        ios: barcode
                    });
                }
                _this.notify({
                    eventName: MLKitBarcodeScanner.scanResultEvent,
                    object: _this,
                    value: result
                });
            }
        };
    };
    MLKitBarcodeScanner.prototype.rotateRecording = function () {
        return false;
    };
    return MLKitBarcodeScanner;
}(barcodescanning_common_1.MLKitBarcodeScanner));
exports.MLKitBarcodeScanner = MLKitBarcodeScanner;
function getBarcodeDetector(formats) {
    if (formats && formats.length > 0) {
        var barcodeFormats_1 = 0;
        formats.forEach(function (format) { return barcodeFormats_1 |= format; });
        return FIRVision.vision().barcodeDetectorWithOptions(FIRVisionBarcodeDetectorOptions.alloc().initWithFormats(barcodeFormats_1));
    }
    else {
        return FIRVision.vision().barcodeDetector();
    }
}
function scanBarcodesOnDevice(options) {
    return new Promise(function (resolve, reject) {
        try {
            var barcodeDetector = getBarcodeDetector(options.formats);
            barcodeDetector.detectInImageCompletion(getImage(options), function (barcodes, error) {
                if (error !== null) {
                    reject(error.localizedDescription);
                }
                else if (barcodes !== null) {
                    var result = {
                        barcodes: []
                    };
                    for (var i = 0, l = barcodes.count; i < l; i++) {
                        var barcode = barcodes.objectAtIndex(i);
                        result.barcodes.push({
                            value: barcode.rawValue,
                            format: barcodescanning_common_1.BarcodeFormat[barcode.format],
                            ios: barcode
                        });
                    }
                    resolve(result);
                }
            });
        }
        catch (ex) {
            console.log("Error in firebase.mlkit.scanBarcodesOnDevice: " + ex);
            reject(ex);
        }
    });
}
exports.scanBarcodesOnDevice = scanBarcodesOnDevice;
function getImage(options) {
    var image = options.image instanceof image_source_1.ImageSource ? options.image.ios : options.image.imageSource.ios;
    return FIRVisionImage.alloc().initWithImage(image);
}
